### PR TITLE
Move nrm2 from FORTRAN to cblas interface

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

File: src/lib_dispatch.cpp [hunk 1]
```diff
@@ -224,31 +225,27 @@ void rotg(P *a, P *b, P *c, P *s, resource const resrc)
...
 #ifdef ASGARD_USE_CUDA
-    // no non-fp blas on device
-    expect(std::is_floating_point_v<P>);
     P norm;
     // function instantiated for these two fp types
     if constexpr (std::is_same<P, double>::value)
     {
-      auto const success =
-          cublasDnrm2(device.get_handle(), *n, x, *incx, &norm);
+      auto const success = cublasDnrm2(device.get_handle(), n, x, incx, &norm);
```
Comment (mkstoyanov) [inline src/lib_dispatch.cpp]: Isn't `device` global? I don't mind (under the assumption of one GPU per MPI rank), I'm just saying.

File: src/lib_dispatch.cpp [hunk 2]
```diff
@@ -224,31 +225,27 @@ void rotg(P *a, P *b, P *c, P *s, resource const resrc)
   }
 }
 
-template<typename P>
-P nrm2(int *n, P *x, int *incx, resource const resrc)
+template<typename P, resource resrc>
```
Comment (mkstoyanov) [inline src/lib_dispatch.cpp]: I don't like the loss of auto-detection for `P`. At least, we can set `resource` to default to host and be explicit only when we want to go to the device.

Comment (quantumsteve) [reply src/lib_dispatch.cpp]: I brought back auto-detection for `P` at the expense of separate functions for host and device implementations.

**Comment by quantumsteve** [comment]
closing in favor of #546

**Pull request closed without merging**